### PR TITLE
Call Lifecycle Hooks On InfinityModel

### PR DESCRIPTION
**Pull request by johncalvinyoung**
This is first-draft spike code to attempt to answer #292. I didn't get the current test harness figured out well enough to write tests for it yet--any tips and I'll be glad to add. I'm trying to preserve current semantics and signatures (route customization still permitted, etc) but add an additional entry point for customization (embedding the hook in the `infinityModel` subclass).

### CODE REVIEW

File: addon/mixins/route.js [hunk 1]
```diff
@@ -161,7 +159,9 @@ const RouteMixin = Mixin.create({
       totalPagesParam,
       _infinityModelName: modelName,
       extraParams: options,
-      content: A()
+      content: A(),
+      _routeAfterInfinityLoad: get(this, 'afterInfinityLoad'),
+      _routeInfinityModelLoaded: get(this, 'infinityModelLoaded')
```
Comment (snewcomer) [inline addon/mixins/route.js]: Would these be public?  (Sry not at my computer and don’t see where hey are used)

Comment (johncalvinyoung) [reply addon/mixins/route.js]: Not intended to be, no. But I didn't specifically document public/private behavior here.

**Comment by hhff** [comment]
I tend to think the ergonomics would be slightly better with a name like `didLoadPage` - the `afterInfinityModel` notation was only that as it was declared on the route.

File: addon/lib/infinity-model.js [hunk 2]
```diff
@@ -179,5 +180,50 @@ export default ArrayProxy.extend({
...
+    @return {Ember.RSVP.Promise} A Promise that resolves the new objects
+    @return {Ember.Array} the new objects
+   */
+  _afterInfinityModel(newObjects, infinityModel) {
+    let result = this.afterInfinityModel(newObjects, infinityModel);
+
+    if(!this._routeAfterInfinityModel || typeof this._routeAfterInfinityModel !== 'function') {
+      return result || newObjects;
+    } else {
+      return result.then((result) => { return this._routeAfterInfinityModel(result || newObjects, infinityModel) || newObjects });
```
Comment (snewcomer) [inline addon/lib/infinity-model.js]: if it fails the first check (`this._routeAfterInfinityModel`), then this still could fail if it didn't exist.

File: addon/lib/infinity-model.js [hunk 3]
```diff
@@ -179,5 +180,50 @@ export default ArrayProxy.extend({
...
+  /**
+    abstract after-model hook, can be overridden in subclasses
+
+    @method afterInfinityModel
+    @param {Ember.Array} newObjects the new objects added to the model
+    @param {Ember.ArrayProxy} infinityModel (self)
+    @return {Ember.RSVP.Promise} A Promise that resolves the new objects
+    @return {Ember.Array} the new objects
+   */
+  afterInfinityModel(newObjects, infinityModel) {
```
Comment (snewcomer) [inline addon/lib/infinity-model.js]: I like this!  backwards compat with the route way as well 👍

File: addon/lib/infinity-model.js [hunk 4]
```diff
@@ -179,5 +180,50 @@ export default ArrayProxy.extend({
...
+      return result || newObjects;
+    } else {
+      return result.then((result) => { return this._routeAfterInfinityModel(result || newObjects, infinityModel) || newObjects });
+    }
+  }
+
+  /**
+    abstract after-model hook, can be overridden in subclasses
+
+    @method afterInfinityModel
```
Comment (snewcomer) [inline addon/lib/infinity-model.js]: mark this with `@public`?

**Comment by johncalvinyoung** [comment]
Thanks for the comments. I'll try to circle back around to this and update the branch in the next couple of days. I'm really tight on open-source time right now.

**Comment by snewcomer** [comment]
@johncalvinyoung Hey! 👋  Lmk if you want any help to push this across the line.  Really appreciate the work so far!

**Comment by snewcomer** [comment]
@johncalvinyoung 👋  Any time this week?  Would like to get 1.0 released this week if it works out for everybody!

**Comment by snewcomer** [comment]
@johncalvinyoung Thank you so much for this PR.  This should be fixed in master (I stole a bunch of your awesome work here :) )

Lmk if you encounter any errors.

**Comment by johncalvinyoung** [comment]
@snewcomer Thanks so much. Sorry for letting this go stale. I'm overcommitted at the moment and my primary project pulled my Phase 2 ticket where I was hoping to work through some of this. I may get to get back to it next month. :/ Glad to hear you're integrating the pattern into master...I'll take a look at it and see if I see any gaps that need to be filled.

**Pull request closed without merging**